posts/2018-05-09-tagless-final-streaming.md

```diff
@@ -204,6 +204,33 @@ While in this case it was possible to make it generic I don't recommend to do th
 
 What I recommend instead, is to write this kind of logic in the streaming interpreter itself. You could also write a generic program that implements the parts that can be abstracted (eg. applying a discount to an item `f: Item => Item`) and leave the other parts to the interpreter.
 
+### Design alternative
+
+Another possible and very interesting alternative suggested by [Michael Pilquist](https://github.com/mpilquist), would be to define our repository as follows:
+
+```scala
+trait ItemRepository[F[_], S[_[_], _]] {
+  def findAll: S[F, Item]
+}
+```
+
+Where the second type parameter matches the shape of `fs2.Stream`. In this case our streaming repository will remain the same (it should just extend `ItemRepository[F, Stream]` instead of `ItemRepository[F, Stream[F, ?]]`) but our in memory interpreter will now rely on `fs2.Stream` instead of a parametric `G[_]`, for example:
+
+```scala
+object MemRepositoryAlt extends ItemRepository[Id, Stream] {
+
+  override def findAll: Stream[Id, Item] = Stream.emits {
+    sql"select name, price from items"
+      .query[Item]
+      .to[List]
```

Comment:
Instead of calling .to[List], call .stream — then no need for the #tream.emits call and everything is streamed from DB.

Reply:
You're absolutely right :D

Reply:
I did it in the `StreamingRepository` and didn't realize about this one :p